src/Propel/Runtime/ActiveQuery/Criterion/AbstractInnerQueryCriterion.php

```diff
@@ -0,0 +1,149 @@
+
+/**
+ * Abstract Criterion for nested filter expression, that bind an inner query
+ * with an operator like IN, EXISTS
+ *
+ * @phpstan-consistent-constructor
+ */
+abstract class AbstractInnerQueryCriterion extends AbstractCriterion
+{
+    /**
+     * @var string|null Left side of the operator, can be empty.
+     */
+    protected $leftOperand;
+
+    /**
+     * @var string|null The sql operator expression, i.e. "IN" or "NOT IN".
+     */
+    protected $sqlOperator;
+
+    /**
+     * @var \Propel\Runtime\ActiveQuery\Criteria
+     */
+    protected $innerQuery;
+
+    /**
+     * Resolves the operator as given by the user to the SQL operator statement.
+     *
+     * @param string $operatorDeclaration
+     *
+     * @return string
+     */
+    abstract protected function resolveOperator(string $operatorDeclaration): string;
+
+    /**
+     * Allows to edit or replace the inner query before it is turned to SQL.
+     *
+     * @return \Propel\Runtime\ActiveQuery\Criteria
+     */
+    abstract protected function processInnerQuery(): Criteria;
+
+    /**
+     * Allows to edit or replace the inner query before it is turned to SQL.
+     *
+     * @param mixed $outerQuery
+     * @param \Propel\Runtime\Map\RelationMap $relationMap
+     * @param string|null $operator
+     * @param \Propel\Runtime\ActiveQuery\ModelCriteria $innerQuery
+     *
+     * @return self
+     */
+    public static function createForRelation($outerQuery, RelationMap $relationMap, ?string $operator, ModelCriteria $innerQuery): self
+    {
+        $filter = new static($outerQuery, null, $operator, $innerQuery);
+        $filter->initForRelation($outerQuery, $relationMap);
+
+        return $filter;
+    }
+
+    /**
+     * @param \Propel\Runtime\ActiveQuery\Criteria $outerQuery
+     * @param string|null $leftOperand Left side of the operator, usually a column name or empty.
+     * @param string|null $operator The operator, like IN, NOT IN, EXISTS, NOT EXISTS
+     * @param \Propel\Runtime\ActiveQuery\Criteria $innerQuery
+     */
+    public function __construct(
+        $outerQuery,
+        ?string $leftOperand,
+        ?string $operator,
+        Criteria $innerQuery
+    ) {
+        parent::__construct($outerQuery, '', null, null);
+
+        if ($operator) {
+            $operator = trim($operator); // Criteria::IN is padded with spaces
+        }
+
+        $this->leftOperand = $leftOperand;
+        $this->sqlOperator = $this->resolveOperator($operator);
+        $this->innerQuery = $innerQuery;
+    }
+
+    /**
+     * Entry point for child classes to add information about the relation to the query.
+     *
+     * @param \Propel\Runtime\ActiveQuery\ModelCriteria $outerQuery
+     * @param \Propel\Runtime\Map\RelationMap $relation
+     *
+     * @return void
+     */
+    protected function initForRelation(ModelCriteria $outerQuery, RelationMap $relation): void
+    {
+        // does nothing per default
+    }
```

Comment:
Maybe we could define it `abstract` instead?

Reply:
Right, thank you, didn't realize that I didn't need a default anymore.